File: tags/NHI1-0.17/theLink/example/c/Filter6.c
/**
 *  \file       theLink/example/c/Filter6.c
 *  \brief      \$Id$
 *  
 *  (C) 2011 - NHI - #1 - Project - Group
 *  
 *  \version    \$Rev$
 *  \author     EMail: aotto1968 at users.berlios.de
 *  \attention  this software has GPL permissions to copy
 *              please contact AUTHORS for additional information
 */

#include "msgque.h"
#include "debug.h"
#include <stdlib.h>
#include <string.h>

#define ARGS  struct MqS * const mqctx, void *data
#define TRANSCTX ((struct FilterCtxS*const)mqctx)
#define MQ_CONTEXT_S mqctx
#define SETUP_ctx struct FilterCtxS*const ctx = TRANSCTX
#define CHECK_ARGS(s) \
  if (MqReadGetNumItems(mqctx))  { \
    return MqErrorV (mqctx, __func__, -1, "usage: %s (%s)\n", __func__, s); \
  }

struct FilterCtxS {
  struct MqS	    mqctx;
  FILE		    *FH;
};

/*****************************************************************************/
/*                                                                           */
/*                               Request Handler                             */
/*                                                                           */
/*****************************************************************************/

static void ErrorWrite (
  struct MqS *mqctx
)
{
  SETUP_ctx;
  fseek(ctx->FH, 0L, SEEK_END);
  fprintf(ctx->FH, "ERROR: %s\n", MqErrorGetText(mqctx));
  fflush(ctx->FH);
  MqErrorReset (mqctx);
}

static enum MqErrorE FilterEvent (
  struct MqS * const mqctx,
  MQ_PTR const data
)
{
  MQ_TRA count;
  MqErrorCheck (MqStorageCount (mqctx, &count));
  // check if an item is available
  if (count == 0) {
    // no data available
    return MqErrorSetCONTINUE(mqctx);
  } else {
    MQ_TRA Id = 0LL;
    struct MqS * ftr;
    MqErrorCheck (MqServiceGetFilter(mqctx, 0, &ftr));

    // if connection is down -> connect again
    MqErrorCheck1 (MqLinkConnect (ftr));

    // fill the read-buffer from storage
    MqErrorCheck2 (MqStorageSelect (mqctx, &Id));

    // send BDY data to the link-target, on error write message but do not stop processing
    if (MqErrorCheckI (MqReadForward(mqctx, ftr, NULL))) {
      if (MqErrorIsEXIT(mqctx)) {
	return MqErrorReset(mqctx);
      } else {
	goto error2;
      }
    }
    MqStorageDelete (mqctx, Id);
    return MQ_OK;
error1:
    MqErrorCopy(mqctx,ftr);
error2:
    ErrorWrite (mqctx);
    if (Id != 0LL) MqStorageDelete (mqctx, Id);
    return MQ_OK;
  }
error:
  return MqErrorStack(mqctx);
}

static enum MqErrorE LOGF ( ARGS ) {
  SETUP_ctx;
  struct MqS * ftr;
  MqErrorCheck(MqServiceGetFilter(mqctx, 0, &ftr));
  if (!strcmp(MqLinkGetTargetIdent (ftr),"transFilter")) {
    MqErrorCheck (MqReadForward(mqctx, ftr, NULL));
  } else {
    MQ_CST file;
    MqErrorCheck (MqReadC (mqctx, &file));
    ctx->FH = fopen (file, "a");
  }
error:
  return MqSendRETURN(mqctx);
}

static enum MqErrorE WRIT ( ARGS ) {
  MQ_CST str;
  struct FilterCtxS *ctx;
  MqErrorCheck (MqServiceGetFilter (mqctx, 0, (struct MqS**) &ctx));
  MqErrorCheck (MqReadC (mqctx, &str));
  fseek(ctx->FH, 0L, SEEK_END);
  fprintf(ctx->FH, "%s\n", str);
  fflush(ctx->FH);
error:
  return MqSendRETURN(mqctx);
}

static enum MqErrorE EXIT ( ARGS ) {
  MqExit(mqctx);
  return MQ_OK;
  //exit(1);
}

static enum MqErrorE FilterIn ( ARGS ) {
  MqErrorCheck (MqStorageInsert (mqctx, NULL));
error:
  return MqSendRETURN(mqctx);
}

/*****************************************************************************/
/*                                                                           */
/*                                context_init                               */
/*                                                                           */
/*****************************************************************************/

static enum MqErrorE
FilterCleanup (
  struct MqS * const mqctx,
  MQ_PTR data
)
{
  SETUP_ctx;
  if (ctx->FH != NULL) fclose(ctx->FH);
  return MQ_OK;
}

static enum MqErrorE
FilterSetup (
  struct MqS * const mqctx,
  MQ_PTR data
)
{
  register SETUP_ctx;
  struct MqS *ftr;
  MqErrorCheck (MqServiceGetFilter (mqctx, 0, &ftr));

  MqErrorCheck (MqServiceCreate	  (mqctx, "LOGF", LOGF, NULL, NULL));
  MqErrorCheck (MqServiceCreate	  (mqctx, "EXIT", EXIT, NULL, NULL));
  MqErrorCheck (MqServiceStorage  (mqctx, "PRNT"));
  MqErrorCheck (MqServiceCreate	  (mqctx, "+ALL", FilterIn, NULL, NULL));
  MqErrorCheck (MqServiceCreate	  (ftr,   "WRIT", WRIT, NULL, NULL));
  MqErrorCheck (MqServiceProxy	  (ftr,   "+TRT", 0));

  ctx->FH = NULL;

error:
  return MqErrorStack(mqctx);
}

enum MqErrorE
Filter6Factory (
  struct MqS * const tmpl,
  enum MqFactoryE create,
  struct MqFactoryS * const item,
  struct MqS **contextP
)
{ 
  struct MqS * const mqctx = *contextP = MqContextCreate(sizeof(struct FilterCtxS),tmpl);
  
  mqctx->setup.Child.fCreate	    = MqLinkDefault;
  mqctx->setup.Parent.fCreate	    = MqLinkDefault;
  mqctx->setup.isServer		    = MQ_YES;
  mqctx->setup.ServerSetup.fCall    = FilterSetup;
  mqctx->setup.ServerCleanup.fCall  = FilterCleanup;
  mqctx->setup.ignoreExit	    = MQ_YES;

  MqConfigSetEvent (mqctx, FilterEvent, NULL, NULL, NULL);
  return MQ_OK;
}

/*****************************************************************************/
/*                                                                           */
/*                                  main                                     */
/*                                                                           */
/*****************************************************************************/

int
main (
  const int argc,
  MQ_CST argv[]
)
{
  // parse the command-line
  MQ_BFL args = MqBufferLCreateArgs (argc, argv);

  // call Factory 
  struct MqS *mqctx = MqFactoryNew (MQ_ERROR_PANIC, NULL,
    MqFactoryAdd(MQ_ERROR_PANIC, "transFilter", Filter6Factory, NULL, NULL, NULL, NULL, NULL, NULL, NULL)
  );

  // create the link
  MqErrorCheck(MqLinkCreate (mqctx, &args));

  // start event-loop and wait forever
  MqProcessEvent (mqctx, MQ_TIMEOUT_USER, MQ_WAIT_FOREVER);

  // finish and exit
error:
  MqExit (mqctx);
}

File: tags/NHI1-0.17/theLink/example/perl/Filter6.pl
#+
#:  \file       theLink/example/perl/Filter6.pl
#:  \brief      \$Id$
#:  
#:  (C) 2011 - NHI - #1 - Project - Group
#:  
#:  \version    \$Rev$
#:  \author     EMail: aotto1968 at users.berlios.de
#:  \attention  this software has GPL permissions to copy
#:              please contact AUTHORS for additional information
#:

use strict;
use Net::PerlMsgque;
use FileHandle;

$| = 1;

package Filter6;
use base qw(Net::PerlMsgque::MqS);

  sub ErrorWrite {
    my $ctx = shift;
    my $FH = $ctx->DictGet("FH");
    if (defined($FH)) {
      print $FH "ERROR: " . $ctx->ErrorGetText() . "\n";
    } else {
      $ctx->Log("ErrorWrite", 0, $ctx->ErrorGetText() . "\n");
    }
    $ctx->ErrorReset();
  }

  sub WRIT {
    my $ftr = shift;
    my $FH = $ftr->ServiceGetFilter()->DictGet("FH");
    if (defined($FH)) {
      print $FH $ftr->ReadC() . "\n";
    } else {
      $ftr->Log("ErrorWrite", 0, $ftr->ReadC() . "\n");
    }
    $ftr->SendRETURN();
  }

  sub EXIT {
    exit 1;
  }

  sub LOGF {
    my $ctx = shift;
    my $ftr = $ctx->ServiceGetFilter();
    if ($ftr->LinkGetTargetIdent() eq "transFilter") {
      $ctx->ReadForward($ftr);
    } else {
      open(my $FH, ">>", $ctx->ReadC());
      $FH->autoflush(1);
      $ctx->DictSet("FH", $FH);
    }
    $ctx->SendRETURN();
  }

  sub Event {
    my $ctx = shift;
    if ($ctx->StorageCount() == 0) {
      $ctx->ErrorSetCONTINUE();
    } else {
      my $id = 0;
      eval {
	my $ftr = $ctx->ServiceGetFilter();
	$ftr->LinkConnect();
	$id = $ctx->StorageSelect();
	$ctx->ReadForward($ftr);
      };
      if ($@) {
	$ctx->ErrorSet($@);
	if ($ctx->ErrorIsEXIT) {
	  $ctx->ErrorReset();
	  return;
	} else {
	  $ctx->ErrorWrite();
	}
      }
      $ctx->StorageDelete($id);
    }
  }

  sub FilterIn {
    my $ctx = shift;
    $ctx->StorageInsert();
    $ctx->SendRETURN();
  }

  sub ServerCleanup {
    my $ctx = shift;
    my $FH = $ctx->DictGet("FH");
    if (defined($FH)) {
      close($FH);
    }
  }

  sub ServerSetup {
    my $ctx = shift;
    my $ftr = $ctx->ServiceGetFilter();
    $ctx->ServiceCreate("+ALL", \&FilterIn);
    $ctx->ServiceStorage("PRNT");
    $ctx->ServiceCreate("LOGF", \&LOGF);
    $ctx->ServiceCreate("EXIT", \&EXIT);
    $ftr->ServiceCreate("WRIT", \&WRIT);
    $ftr->ServiceProxy("+TRT");
    $ctx->DictSet("FH", undef);

  }

  sub new {
    my $class = shift;
    my $ctx = $class->SUPER::new(@_);
    $ctx->ConfigSetIgnoreExit(1);
    $ctx->ConfigSetServerSetup(\&ServerSetup);
    $ctx->ConfigSetServerCleanup(\&ServerCleanup);
    $ctx->ConfigSetEvent(\&Event);
    bless $ctx, $class;
    return $ctx;
  }


package main;

  our $srv = MqFactoryS::Add("transFilter", "Filter6")->New();
  eval {
    $srv->LinkCreate(@ARGV);
    $srv->ProcessEvent(Net::PerlMsgque::WAIT_FOREVER);
  };
  if ($@) {
    $srv->ErrorSet($@);
  }
  $srv->Exit();

1;
